### DISCUSSION TITLE
Cloudflare `Workers` / `Pages` 自定义域名使用 `自选 IP` 的简单教程

### DISCUSSION THREAD

**Discussion by XIU2** [XIU2/CloudflareSpeedTest#310] [分享]
Cloudflare Workers 自定义域名使用自选 IP 其实很简单，只需要做一些小改动即可。

****

### Workers 自定义域名 正常情况下的常规步骤：

1. 正常部署完 Workers 后，
2. 添加一个**子域名**记录（指向 IP 随意）并开启代理（橙色 ☁），
3. 去该域名的 Workers 侧栏选项中**添加路由**将刚刚添加的 子域名 指向部署的 Workers 服务。
4. 配置完成等待生效后，手动访问一次刚刚添加的自定义域名，确保正常

****

### 指定自选 IP 的小改动：

想要自选 IP，那么只需要在添加子域名记录的时候，**不开启代理（橙色 ☁）**，直接指向**自选 IP** 即可。

**Comment by XIU2** [comment]
> [!Warning]
> 注意！可能是因为这样自选的人太多了？导致 Cloudflare 限制了下面教程方案，如果你的域名解析托管在 Cloudflare CDN，那么会被自动添加 CNAME 解析记录并无法修改，但如果你的域名解析没有托管在 Cloudflare CDN，那就不受影响，依然可以随意修改。

****

**Cloudflare Pages** 自定义域名使用自选 IP 和 Workers 差不多，也是只需要做一些小改动即可。

****

### Pages 自定义域名 正常情况下的常规步骤：

1. 正常部署完 Pages 后，
2. 点进去 Pages 项目，选择顶部 `自定义域` 选项卡，点击右边的 `设置自定义域`，
3. 填写自定义域名，配置域名 CNAME 解析记录（如果域名就托管在 Cloudflare 那么会自动完成，反之则需要手动添加），
4. 配置完成等待生效后，手动访问一次刚刚添加的自定义域名，确保正常

****

### 指定自选 IP 的小改动：

在确保自定义域名能正常访问后（即已签发 1 年 SSL 证书），就可以修改域名解析记录了。

如楼下所说，专门用一个子域名 A 记录解析指向自选 IP（不开橙色 ☁），然后修改旧 CNAME 的解析记录目标为该子域名（不开橙色 ☁）

> 注意：如果你的域名托管在 Cloudflare，那么因为后续 Cloudflare 加了自动检查机制（以前是没有的），会导致指向非指定 CNAME 时被自动停用（Error 1000），因此请把域名托管到其他地方，再去修改 CNAME 解析记录目标。

**Reply by domoxiaojun** [reply]
自定义域名解析后,没办法删除CNAME了

**Reply by m0rml1n** [reply]
请问这个方法还有用吗？

**Reply by m0rml1n** [reply]
Error 1000
DNS points to prohibited IP

**Reply by XIU2** [reply]
@m0rml1n 经过测试，我目前的几个 Workers / Pages 网站都可以正常访问（从我写这篇教程之前，我就指向自选 IP 了）。
你用不了的话，要么是**使用姿势有问题**，要么是**指向的 IP 被限制**，要么是 Cloudflare 改了什么机制但又不影响我这种以前干的。

**Reply by m0rml1n** [reply]
@XIU2 感谢回复，我这里确保自定义域名能正常访问后，删除 CNAME解析记录，添加A 解析记录（不开橙色 ☁）。就Error 1000了，去 cloudflare pages 里，看到的是 域名 已被禁用，应该是 CNAME 失效就被禁用了

**Reply by XIU2** [reply]
@m0rml1n 你尝试先搞个子域名 A 记录指向自选 IP（不开橙色 ☁），例如 `cf.xxx.xxx`，然后把自定义域名自动添加的 CNAME 解析记录目标改成你的域名（即 `cf.xxx.xxx`，依然不开橙色 ☁）。

为了避免其他因素干扰，在这样做之前，请先删除你的 Pages 项目，并删除刚才你添加的各解析记录，重新创建 Pages 项目并添加自定义域名，然后再去按上面说的去做。

> 我的多个 Workers / Pages 网站为了方便管理也是如楼下所说专门整个子域名用于 CNAME 指向的。  

> 上面这样操作后，如果正常了，那么说明 Cloudflare 仅检查是否有 CNAME 记录而并没有检查指向的目标是什么。  
> 如果还不行，那可能是官方封住了这些路。

**Reply by m0rml1n** [reply]
@XIU2 感谢，找到原因了，之前我的域名dns解析在cf，老失败，应该就是这个原因。后来放到别的地方，就可以了，这个方法还是有用

**Comment by XIU2** [comment]
如果你同时有多个 Workers / Pages 自定义域名，那么为了方便管理 及 更新自选 IP，建议专门添加一个用来 **指向自选 IP 的 A 记录子域名**，然后让那些 Workers / Pages 自定义域名都 **CNAME 解析到该子域名**即可。

后续只要更新该子域名指向的**自选 IP**，其他所有 CNAME 指向该子域名的自定义域名也**都会跟着更新**了。

``` yaml
# 这样的话，用户访问你的自定义域名，DNS 解析流程就是这样的了：
访客  =[DNS 解析]=>  自定义域名  =[CANME 指向]=>  专门子域名  =[A 指向]=> 自选 IP
```

> 例如，专门用来指向自选 IP 的 A 记录子域名为 `cf.aaa.com`，那么其他所有域名都可以 CNAME 指向 `cf.aaa.com`，而你只需要定期更新 `cf.aaa.com` 指向的自选 IP 即可（无论是手动还是通过 Cloudflare API 自动）。

我以前写的一个 Cloudflare API 手动教程及 bat/sh 脚本 #40 ，可以很方便的 **自动更新域名指向的自选 IP**。

**Comment by banshicy** [comment]
大佬们，这样似乎不行，CF会提示禁止指向CF的IP地址

**Comment by XIU2** [comment]
@banshicy Cloudflare CDN 的一些 IP 是**限制免费账户（的域名）使用**的，比如 `1.1.1.1`，你通过这些 IP 访问你的域名（不管你是 Hosts 指定，还是域名解析为自选 IP，本质是一样的），就会提示错误，请更换其他 IP。

目前我自己的 Workers、Pages 用的都是自选 IP。

**Comment by cheyuze** [comment]
大佬，用这种方法做完了，域名也可以访问（自己的域名，访问后会直接下载1gb的测试文件）

其中部署时使用的是这个↓
[http://speedtest-sfo3.digitalocean.com/1gb.test
DigitalOcean 旧金山 1GB，文件反代时注意这是个 http 协议]

但是添加到脚本中（-url https://*****.asia/1gb.test）会出现下载测试0MB/S的情况。延迟正常显示。请问可能的原因有是什么呀？
主要用途是PT做种，访问PT站..第一次接触到这些，努力学习中..

**Comment by XIU2** [comment]
@cheyuze 
这个教程是给 Cloudflare 中的 Workers、Pages 项目强制解析为自选 IP 的，如果你要自建下载测速地址，则和该教程完全无关。

**Comment by cheyuze** [comment]
从https://github.com/XIU2/CloudflareSpeedTest/issues/168 被拉过来的TAT自定义域名
https://github.com/XIU2/CloudflareSpeedTest/issues/168#issuecomment-1028697209

_最后尝试浏览器访问 https://XXX.YYY.workers.dev/200mb.test 如果可以正常下载则代表部署成功
注意：因为 *.workers.dev 域名被 DNS 污染，所以默认分配的子域名将无法使用，有域名的可以将 Workers 服务路由绑定至自己的域名就能访问了。

具体可以看看 https://github.com/XIU2/CloudflareSpeedTest/issues/189 这里提到了 Workers 自定义域名。

后续使用 CloudflareST 时，记得加上 -url 参数，例如：_

**Comment by WincerChan** [comment]
现在好像不行了。去年我按照教程成功过，现在按照这样添加的新域名就不生效了。访问直接报 403

**Reply by yushiro** [reply]
除了添加自定义域名，还需要加个路由转发规则，我近期刚给worker添加优选ip，成功。

**Reply by WincerChan** [reply]
@yushiro 请问 pages 如何添加路由转发规则呢

**Reply by WincerChan** [reply]
我又重新添加了一下，刚添加的时候是可以的，但是过了 1 分钟后，会到 403 页面

**Comment by aihoom** [comment]
我根据这个方式添加了，但是提示重定向过多。小黄朵关闭了

**Comment by superen7527** [comment]
原来是这样 年初一改不是提示不能指向CF 要不就是一直刷新重新试 好不容易手动好了不是1000就是无限重定向 有时候关闭强制s 最后能访问也会出现跳转到别的网址（记不清是用的cname别人的域名这个肯定会还是指定ip也会）而且有的人可以一直加ipA记录我当时只能加一个

**Comment by superen7527** [comment]
现在好像设置的自定义域，如果先在cf指向记录，就无法添加自定义域。如果先添加自定义域，就无法再修改解析记录。

**Reply by onpix** [reply]
我也是一样的问题。。。cloudflare pages要绑定域名，首先会要求把dns nameserver改成cloudflare，但是这样一来就会报1000的错误了。。。

**Comment by xuanxuan1231** [comment]
成功了，搭了个jsdelivr的镜像，好像挺好用的

**Reply by xuanxuan1231** [reply]
要先用`自定义域`添加，等cf给你要到证书之后再解析到自选IP上

**Reply by xuanxuan1231** [reply]
填`自定义路由`的时候要加`*`

**Reply by xuanxuan1231** [reply]
如果没有路由的话就会报1000  
注意`example.com/gh/*`只能管：`example.com/gh/example`  
需要`example.com/gh`才能保证`example.com/gh`不报1000

**Reply by xuanxuan1231** [reply]
（今天刚弄等3个月后再来看看证书

**Comment by hehaokunzhenshuai** [comment]
<img width="1126" alt="图片" src="https://github.com/XIU2/CloudflareSpeedTest/assets/60168969/22340254-abfe-4eb8-a838-075234f1ba47">
启动exe，配置上了自选ip，但是无法访问，麻烦帮忙看下
<img width="1335" alt="图片" src="https://github.com/XIU2/CloudflareSpeedTest/assets/60168969/388acd86-2ac1-4e09-b6ab-f8a687382e08">

**Reply by XIU2** [reply]
该帖子教程我记得已经失效了，当初我公开教程后传播太广，这样干的人太多，导致 Cloudflare 现在加了很多措施来反制。。。
